src/RequestHandler/inprocessapplication.cpp

```diff
@@ -1,21 +1,653 @@
+        }
+
+        m_fRecycleProcessCalled = TRUE;
+
+        // First call into the managed server and shutdown
+        if (m_ShutdownHandler != NULL)
+        {
+            m_ShutdownHandler(m_ShutdownHandlerContext);
+            m_ShutdownHandler = NULL;
+        }
+
+        if (m_hThread != NULL &&
+            GetExitCodeThread(m_hThread, &dwThreadStatus) != 0 &&
+            dwThreadStatus == STILL_ACTIVE)
+        {
+            // wait for gracefullshut down, i.e., the exit of the background thread or timeout
+            if (WaitForSingleObject(m_hThread, dwTimeout) != WAIT_OBJECT_0)
+            {
+                // if the thread is still running, we need kill it first before exit to avoid AV
+                if (GetExitCodeThread(m_hThread, &dwThreadStatus) != 0 && dwThreadStatus == STILL_ACTIVE)
+                {
+                    TerminateThread(m_hThread, STATUS_CONTROL_C_EXIT);
+                }
+            }
+        }
+
+        CloseHandle(m_hThread);
+        m_hThread = NULL;
+        s_Application = NULL;
+
+        ReleaseSRWLockExclusive(&m_srwLock);
+        if (m_pHttpServer && m_pHttpServer->IsCommandLineLaunch())
+        {
+            // IISExpress scenario
+            // Can only call exit to terminate current process
+            exit(0);
+        }
+    }
+}
+
+REQUEST_NOTIFICATION_STATUS
+IN_PROCESS_APPLICATION::OnAsyncCompletion(
+    IHttpContext* pHttpContext,
+    DWORD           cbCompletion,
+    HRESULT         hrCompletionStatus
+)
+{
+    HRESULT hr;
+    IN_PROCESS_STORED_CONTEXT* pInProcessStoredContext = NULL;
+    REQUEST_NOTIFICATION_STATUS dwRequestNotificationStatus = RQ_NOTIFICATION_CONTINUE;
+
+    hr = IN_PROCESS_STORED_CONTEXT::GetInProcessStoredContext(pHttpContext, &pInProcessStoredContext);
+    if (FAILED(hr))
+    {
+        // Finish the request as we couldn't get the callback
+        pHttpContext->GetResponse()->SetStatus(500, "Internal Server Error", 19, hr);
+        return RQ_NOTIFICATION_FINISH_REQUEST;
+    }
+    else if (pInProcessStoredContext->QueryIsManagedRequestComplete())
+    {
+        // means PostCompletion has been called and this is the associated callback.
+        dwRequestNotificationStatus = pInProcessStoredContext->QueryAsyncCompletionStatus();
+        // TODO cleanup whatever disconnect listener there is
+        return dwRequestNotificationStatus;
+    }
+    else
+    {
+        // Call the managed handler for async completion.
+        return m_AsyncCompletionHandler(pInProcessStoredContext->QueryManagedHttpContext(), hrCompletionStatus, cbCompletion);
+    }
+}
+
+REQUEST_NOTIFICATION_STATUS
+IN_PROCESS_APPLICATION::OnExecuteRequest(
+    _In_ IHttpContext* pHttpContext
+)
+{
+    if (m_RequestHandler != NULL)
+    {
+        return m_RequestHandler(pHttpContext, m_RequstHandlerContext);
+    }
+
+    ////
+    //// return error as the application did not register callback
```

Comment:
We need to centralize all FREB/log message to somewhere besides the forwarding handler.